azblob/zt_examples_test.go

```diff
@@ -1302,3 +1304,244 @@ func ExampleListBlobsHierarchy() {
+//	}
+//	defer func() { // resp and Body should not be nil
+//		io.Copy(ioutil.Discard, resp.Body)
+//		resp.Body.Close()
+//	}()
+//
+//	// Check if the status code indicates success
+//	// The request returns 200 currently, add 201 and 202 as well for possible extension.
+//	if !(HTTPResponseExtension{Response: resp}).IsSuccessStatusCode(http.StatusOK, http.StatusCreated, http.StatusAccepted) {
+//		return nil, fmt.Errorf("failed to get token from msi, status code: %v", resp.StatusCode)
+//	}
+//
+//	b, err := ioutil.ReadAll(resp.Body)
+//	if err != nil {
+//		return nil, err
+//	}
+//
+//	result := &adal.Token{}
+//	if len(b) > 0 {
+//		b = ByteSliceExtension{ByteSlice: b}.RemoveBOM()
+//		if err := json.Unmarshal(b, result); err != nil {
+//			return nil, fmt.Errorf("failed to unmarshal response body, %v", err)
+//		}
+//	} else {
+//		return nil, errors.New("failed to get token from msi")
+//	}
+//
+//	return result, nil
+//}
+// //==================================================================================================================================
+
+func fetchMSIToken(applicationID string, identityResourceID string, resource string, callbacks ...adal.TokenRefreshCallback) (*adal.ServicePrincipalToken, error) {
+	// Both application id and identityResourceId cannot be present at the same time.
+	if applicationID != "" && identityResourceID != "" {
+		return nil, fmt.Errorf("didn't expect applicationID and identityResourceID at same time")
+	}
+
+	// msiEndpoint is the well known endpoint for getting MSI authentications tokens
+	// msiEndpoint := "http://169.254.169.254/metadata/identity/oauth2/token" for production Jobs
+	msiEndpoint, _ := adal.GetMSIVMEndpoint()
+
+	var spt *adal.ServicePrincipalToken
+	var err error
+
+	// both can be empty, systemAssignedMSI scenario
+	if applicationID == "" && identityResourceID == "" {
+		spt, err = adal.NewServicePrincipalTokenFromMSI(msiEndpoint, resource, callbacks...)
+	}
+
+	// msi login with clientID
+	if applicationID != "" {
+		spt, err = adal.NewServicePrincipalTokenFromMSIWithUserAssignedID(msiEndpoint, resource, applicationID, callbacks...)
+	}
+
+	// msi login with resourceID
+	if identityResourceID != "" {
+		spt, err = adal.NewServicePrincipalTokenFromMSIWithIdentityResourceID(msiEndpoint, resource, identityResourceID, callbacks...)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return spt, spt.Refresh()
+}
+
+func getOAuthToken(applicationID, identityResourceID, resource string, callbacks ...adal.TokenRefreshCallback) (*TokenCredential, error) {
+	spt, err := fetchMSIToken(applicationID, identityResourceID, resource, callbacks...)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Refresh obtains a fresh token for the Service Principal.
+	err = spt.Refresh()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	tc := NewTokenCredential(spt.Token().AccessToken, func(tc TokenCredential) time.Duration {
+		_ = spt.Refresh()
+		return time.Until(spt.Token().Expires())
```

Comment:
Perhaps it's best to deduct a slight buffer from the expiration so that the token credential is always valid.